### PR TITLE
add fast rsqrt comparison in c++

### PR DESCRIPTION
**Pull request by lu1and10**
This PR add approximate rsqrt calculation in c++. By using Newton iterations in double precision, one can achieve significant  speed up compared to native simd sqrt function.

### CODE REVIEW

File: lap3dkernel/c++SIMD/lap3dkernel.cpp [hunk 1]
```diff
@@ -155,11 +234,26 @@ int main (void)
   Real t = omp_get_wtime()-ts;
   Real ans = 0;
   for(auto& x : pot) ans+=x;    // this answer checks it agrees btw methods, to 15 digits
-  std::cout << std::setprecision(15);
+  std::cout << std::setprecision(16);
```
Comment (lu1and10) [inline lap3dkernel/c++SIMD/lap3dkernel.cpp:237]: the result may differ in 15th digit with rounding. If print to 16 digits, 15th digit match, 16th digit does not match; so I propose change to print 16 digits.

File: lap3dkernel/c++SIMD/VCL_AgnerFog_version1/vectorf256.h [hunk 2]
```diff
@@ -1785,6 +1785,11 @@ static inline Vec4d sqrt(Vec4d const & a) {
     return _mm256_sqrt_pd(a);
 }
 
+// aprrox 1/sqrt(a) function, takes double type as function argument
```
Comment (lu1and10) [inline lap3dkernel/c++SIMD/VCL_AgnerFog_version1/vectorf256.h:1788]: VCL version 1 or version 2 don't have simd approx rsqrt for double type argument, so I added to the local VCL lib.

File: lap3dkernel/c++SIMD/VCL_AgnerFog_version1/vectorf512.h [hunk 3]
```diff
@@ -1379,6 +1379,11 @@ static inline Vec8d sqrt(Vec8d const & a) {
     return _mm512_sqrt_pd(a);
 }
 
+// aprrox 1/sqrt(a) function, takes double type as function argument
```
Comment (lu1and10) [inline lap3dkernel/c++SIMD/VCL_AgnerFog_version1/vectorf512.h:1382]: VCL version 1 or version 2 don't have simd approx rsqrt for double type argument, so I added to the local VCL lib.

File: lap3dkernel/c++SIMD/VCL_AgnerFog_version1/vectorf512e.h [hunk 4]
```diff
@@ -1346,6 +1346,11 @@ static inline Vec8d sqrt(Vec8d const & a) {
     return Vec8d(sqrt(a.get_low()), sqrt(a.get_high()));
 }
 
+// aprrox 1/sqrt(a) function, takes double type as function argument
```
Comment (lu1and10) [inline lap3dkernel/c++SIMD/VCL_AgnerFog_version1/vectorf512e.h:1349]: VCL version 1 or version 2 don't have simd approx rsqrt for double type argument, so I added to the local VCL lib.

File: lap3dkernel/c++SIMD/lap3dkernel.cpp [hunk 5]
```diff
@@ -25,6 +25,7 @@
 
 // choose double prec throughout
 typedef double Real;
+typedef long Integer;
```
Comment (lu1and10) [inline lap3dkernel/c++SIMD/lap3dkernel.cpp:28]: This line should be removed. Integer is not used anywhere. It was used in pow function, but I removed the pow function in latest commit. Should find this before create the PR, sorry. 😐

File: lap3dkernel/c++SIMD/lap3dkernel.cpp [hunk 6]
```diff
@@ -101,6 +102,45 @@ void test_kernel_vec512(const Real* Xs, const Real* Ys, const Real* Zs,
...
+            Vec8d rinv = approx_rsqrt(r2);
+            // Note the trick here changes the Newton iteration constants
+            // the standard two Newton iterations are,
+            // rinv *= ((3.0) - r2 * rinv * rinv) * 0.5;
+            // rinv *= ((3.0) - r2 * rinv * rinv) * 0.5;
+            // we can save one simd multiplication " * 0.5 " by using the following trick
+            // first Newton iteration
+            rinv *= ((3.0) - r2 * rinv * rinv);
+            // second Newton iteration
+            rinv *= ((12.0) - r2 * rinv * rinv) * (0.0625);
```
Comment (lu1and10) [inline lap3dkernel/c++SIMD/lap3dkernel.cpp:134]: The modified Newton iteration saves one simd multiplication, the standard Newton iteration should be 
rinv *= ((3.0) - r2 * rinv * rinv) * 0.5

File: lap3dkernel/c++SIMD/lap3dkernel.cpp [hunk 7]
```diff
@@ -126,6 +166,45 @@ void test_kernel_vec256(const Real* Xs, const Real* Ys, const Real* Zs,
...
+            Vec4d rinv = approx_rsqrt(r2);
+            // Note the trick here changes the Newton iteration constants
+            // the standard two Newton iterations are,
+            // rinv *= ((3.0) - r2 * rinv * rinv) * 0.5;
+            // rinv *= ((3.0) - r2 * rinv * rinv) * 0.5;
+            // we can save one simd multiplication " * 0.5 " by using the following trick
+            // first Newton iteration
+            rinv *= ((3.0) - r2 * rinv * rinv);
+            // second Newton iteration
+            rinv *= ((12.0) - r2 * rinv * rinv) * (0.0625);
```
Comment (lu1and10) [inline lap3dkernel/c++SIMD/lap3dkernel.cpp:199]: The modified Newton iteration saves one simd multiplication, the standard Newton iteration should be
rinv *= ((3.0) - r2 * rinv * rinv) * 0.5

**Comment by ahbarnett** [comment]
Thanks!
I presented results so far in group meeting, caused some discussion. (We are not telling people what language to choose!) Also intel fortran may have fast rsqrt too.
Talk soon, Alex

**Pull request merged**